drivers/soundwire/sysfs_slave_dpn.c

```diff
@@ -0,0 +1,348 @@
+sdw_dpn_attr(min_word, "%d\n");
+sdw_dpn_attr(type, "%d\n");
+sdw_dpn_attr(max_grouping, "%d\n");
+sdw_dpn_attr(simple_ch_prep_sm, "%d\n");
+sdw_dpn_attr(ch_prep_timeout, "%d\n");
+sdw_dpn_attr(max_ch, "%d\n");
+sdw_dpn_attr(min_ch, "%d\n");
+sdw_dpn_attr(max_async_buffer, "%d\n");
+sdw_dpn_attr(block_pack_mode, "%d\n");
+sdw_dpn_attr(port_encoding, "%d\n");
+
+#define sdw_dpn_array_attr(field, format_string)			\
+									\
+static ssize_t field##_show(struct sdw_slave *slave,			\
+			    int N,					\
+			    int dir,					\
+			    struct dpn_attribute *attr,			\
+			    char *buf)					\
+{									\
+	struct sdw_dpn_prop *dpn;					\
+	unsigned long mask;						\
+	ssize_t size = 0;						\
+	int bit;							\
+	int i;								\
+	int j;								\
+									\
+	if (dir) {							\
+		dpn = slave->prop.src_dpn_prop;				\
+		mask = slave->prop.source_ports;			\
+	} else {							\
+		dpn = slave->prop.sink_dpn_prop;			\
+		mask = slave->prop.sink_ports;				\
+	}								\
+									\
+	i = 0;								\
+	for_each_set_bit(bit, &mask, 32) {				\
+		if (bit == N) {						\
+			for (j = 0; j < dpn[i].num_##field; j++)	\
+				size += sprintf(buf + size,		\
+						format_string,		\
+						dpn[i].field[j]);	\
+			size += sprintf(buf + size, "\n");		\
+			return size;					\
+		}							\
+		i++;							\
+	}								\
+	return -EINVAL;							\
+}									\
+									\
+static struct dpn_attribute dpn_attr_##field = __ATTR_RO(field)
+
+sdw_dpn_array_attr(words, "%d ");
+sdw_dpn_array_attr(ch_combinations, "%d ");
+sdw_dpn_array_attr(channels, "%d ");
+
+static struct attribute *dpn_attrs[] = {
+	&dpn_attr_max_word.attr,
+	&dpn_attr_min_word.attr,
+	&dpn_attr_words.attr,
+	&dpn_attr_type.attr,
+	&dpn_attr_max_grouping.attr,
+	&dpn_attr_simple_ch_prep_sm.attr,
+	&dpn_attr_ch_prep_timeout.attr,
+	&dpn_attr_imp_def_interrupts.attr,
+	&dpn_attr_min_ch.attr,
+	&dpn_attr_max_ch.attr,
+	&dpn_attr_channels.attr,
+	&dpn_attr_ch_combinations.attr,
+	&dpn_attr_max_async_buffer.attr,
+	&dpn_attr_block_pack_mode.attr,
+	&dpn_attr_port_encoding.attr,
+	NULL,
+};
+
+static const struct attribute_group dpn_group = {
+	.attrs = dpn_attrs,
+};
+
+static void free_dpn_node(struct kobject *kobj,
+			  const struct attribute_group *group)
+{
+	if (kobj) {
+		sysfs_remove_group(kobj, group);
+		kobject_put(kobj);
+	}
+}
+
+static void free_dpn_tree(struct sdw_slave *slave)
+{
+	struct sdw_dpn_tree *tree = slave->dpn_tree;
+	struct kobject **p;
+
+	if (!tree)
+		return;
+	if (tree->nodes) {
+		for (p = tree->nodes; *p; p++)
+			free_dpn_node(*p, &dpn_group);
+		kfree(tree->nodes);
+	}
+	kfree(tree);
+	slave->dpn_tree = NULL;
+}
+
+static int add_dpn_node(struct kobject *parent, int N, int dir,
+			const struct attribute_group *group,
+			struct kobject **res)
+{
+	struct kobject *kobj = kzalloc(sizeof(*kobj), GFP_KERNEL);
+	int err;
+
+	if (!kobj)
+		return -ENOMEM;
+	kobject_init(kobj, &dpn_ktype);
+	err = kobject_add(kobj, parent, "dp:%d:%s", N, dir ? "src" : "sink");
+	if (err < 0)
```

Comment:
aren't we leaking `kobj` here?

Reply:
no, this is freed in free_dpn_node, but there's another problem that that routine will also called sysfs_remove_group which hasn't been created if we exit. So yes it's probably better to call kobject_put() here so that we don't release something that wasn't allocated.